### ISSUE TITLE
-O always downloads the full file

### ISSUE THREAD

**Issue by mruprich** [rockdaboot/wget2#356]
When using -c and -O, the target file is always downloaded in full size. The content-range starts with 0B for every file. Reproducer:
```
# wget -c -O kernel-modules.rpm https://mirror.karneval.cz/pub/linux/fedora/linux/development/41/Everything/x86_64/os/Packages/k/kernel-debug-modules-6.11.4-301.fc41.x86_64.rpm
kernel-modules.rpm   100% [====================================================================>]  106.55M   32.00MB/s
                          [Files: 1  Bytes: 106.55M [23.06MB/s] Redirects: 0  Todo: 0  Errors: 0]
# ls -lh kernel-modules.rpm
-rw-r--r--. 1 root root 107M Oct 20 13:01 kernel-modules.rpm
# truncate --size 50M kernel-modules.rpm
# ls -lh kernel-modules.rpm
-rw-r--r--. 1 root root 50M Nov 18 07:58 kernel-modules.rpm
# wget --debug -c -O kernel-modules.rpm https://mirror.karneval.cz/pub/linux/fedora/linux/development/41/Everything/x86_64/os/Packages/k/kernel-debug-modules-6.11.4-301.fc41.x86_64.rpm
....
18.075855.798 content-range: bytes 0-111736168/111736169   <---- Always range from 0
18.075855.798 [FRAME 1] < HEADERS
kernel-modules.rpm    96% [=================================================================>   ]  103.19M   26.40MB/s
18.075855.798 [FRAME 0] > SETTINGS
kernel-modules.rpm   100% [====================================================================>]  106.55M   27.43MB/s
18.075900.377   ##  response status 206
18.075900.377 keep_alive=1
18.075900.377 _host_remove_job: 0x559e9ef57cf0
18.075900.377 host_remove_job: qsize=0 host->qsize=0
18.075900.377 [0] action=1 pending=0 host=0x559e9ef57690
18.075900.377 closing connection
18.075900.377 [0] action=1 pending=0 host=0x0
18.075900.377 main: wake up
18.075900.378 main: done
18.075900.378 queue_size: [Files: 0  Bytes: 106.55M [20.85MB/s] Redirects: 0  Todo: 0  Errors: 1]
18.075900.378 blacklist https://mirror.karneval.cz/pub/linux/fedora/linux/development/41/Everything/x86_64/os/Packages/k/kernel-debug-modules-6.11.4-301.fc41.x86_64.rpm
```
I narrowed it down to line 3496 in options.c:
```
// truncate output document
  if (config.output_document && strcmp(config.output_document, "-") && !config.dont_write) {
    if (config.unlink) {
      unlink(config.output_document);
    } else {
      int fd = open(config.output_document, O_WRONLY | O_TRUNC | O_BINARY);   <--------
  
      if (fd != -1)
        close(fd);
    } 
  }   
```
Because of O_TRUNC the file is truncated to 0B every time and then in http_create_request in wget.c:
```
if (config.continue_download || config.start_pos || (config.timestamping && config.if_modified_since)) {
    const char *local_filename = config.output_document ? config.output_document : job->blacklist_entry->local_filename;

    /* We never want to continue the robots job. Always grab a fresh copy
     * from the server. */
    if (job->robotstxt == true) {
      unlink(local_filename);
    }

    if (config.continue_download) {
      long long file_size = get_file_size(local_filename);   <----- stat always returns 0
      if (file_size >= 0)
        wget_http_add_header_printf(req, "Range", "bytes=%lld-", file_size);   <---- range will be 0-
    }
```
I was not sure whether the O_TRUNC is intentional or if it is a bug so I did not create a PR yet. Could you please take a look?

Regards,
Michal Ruprich

**Comment by mruprich** [comment]
wget-1.21.4:
```
# wget -O kernel-modules.rpm https://mirror.karneval.cz/pub/linux/fedora/linux/development/41/Everything/x86_64/os/Packages/k/kernel-debug-modules-6.11.4-301.fc41.x86_64.rpm
Resolving mirror.karneval.cz (mirror.karneval.cz)... 2a02:8301:0:2::150, 89.102.0.150
Connecting to mirror.karneval.cz (mirror.karneval.cz)|2a02:8301:0:2::150|:443... failed: No route to host.
Connecting to mirror.karneval.cz (mirror.karneval.cz)|89.102.0.150|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 111736169 (107M) [application/x-redhat-package-manager]
Saving to: ‘kernel-modules.rpm’

kernel-modules.rpm            100%[================================================>] 106.56M  31.9MB/s    in 4.1s    

2024-11-18 11:02:26 (26.1 MB/s) - ‘kernel-modules.rpm’ saved [111736169/111736169]
# ls -lh kernel-modules.rpm
-rw-r--r--. 1 root root 107M Oct 20 13:01 kernel-modules.rpm
# wget -c -O kernel-modules.rpm https://mirror.karneval.cz/pub/linux/fedora/linux/development/41/Everything/x86_64/os/Packages/k/kernel-debug-modules-6.11.4-301.fc41.x86_64.rpm
--2024-11-18 11:02:50--  https://mirror.karneval.cz/pub/linux/fedora/linux/development/41/Everything/x86_64/os/Packages/k/kernel-debug-modules-6.11.4-301.fc41.x86_64.rpm
Resolving mirror.karneval.cz (mirror.karneval.cz)... 2a02:8301:0:2::150, 89.102.0.150
Connecting to mirror.karneval.cz (mirror.karneval.cz)|2a02:8301:0:2::150|:443... failed: No route to host.
Connecting to mirror.karneval.cz (mirror.karneval.cz)|89.102.0.150|:443... connected.
HTTP request sent, awaiting response... 416 Requested Range Not Satisfiable

    The file is already fully retrieved; nothing to do.

# truncate --size 50M kernel-modules.rpm 
# wget -c -O kernel-modules.rpm https://mirror.karneval.cz/pub/linux/fedora/linux/development/41/Everything/x86_64/os/Packages/k/kernel-debug-modules-6.11.4-301.fc41.x86_64.rpm
--2024-11-18 11:03:12--  https://mirror.karneval.cz/pub/linux/fedora/linux/development/41/Everything/x86_64/os/Packages/k/kernel-debug-modules-6.11.4-301.fc41.x86_64.rpm
Resolving mirror.karneval.cz (mirror.karneval.cz)... 2a02:8301:0:2::150, 89.102.0.150
Connecting to mirror.karneval.cz (mirror.karneval.cz)|2a02:8301:0:2::150|:443... failed: No route to host.
Connecting to mirror.karneval.cz (mirror.karneval.cz)|89.102.0.150|:443... connected.
HTTP request sent, awaiting response... 206 Partial Content
Length: 111736169 (107M), 59307369 (57M) remaining [application/x-redhat-package-manager]
Saving to: ‘kernel-modules.rpm’

kernel-modules.rpm            100%[++++++++++++++++++++++==========================>] 106.56M  21.4MB/s    in 2.6s    

2024-11-18 11:03:15 (21.4 MB/s) - ‘kernel-modules.rpm’ saved [111736169/111736169]
```

So old wget can do this so I am assuming this is a bug in wget2? I will send a PR.

**Comment by mruprich** [comment]
Please, take a look at the PR, this was the simplest way I could think of.

**Comment by rockdaboot** [comment]
Thank you, has been fixed by #357 (manually merged).

**Comment by mruprich** [comment]
Great, thanks :)
